### PR TITLE
Improved marshaling of interface "Marshaler"

### PR DESCRIPTION
**Pull request by c0va23**
The Marshal did not work for pointers to interface Marshaller. I added the test cases, which had experienced this problem.

### CODE REVIEW

**Comment by c0va23** [comment]
I do rabase of this PR.

File: AUTHORS [hunk 1]
```diff
@@ -39,4 +39,3 @@ Dean Elbaz <elbaz.dean@gmail.com>
 Mike Berman <evencode@gmail.com>
 Dmitriy Fedorenko <c0va23@gmail.com>
 Zach Marcantel <zmarcantel@gmail.com>
-James Maloney <jamessagan@gmail.com>
```
Comment (0x6e6562) [inline AUTHORS]: Why is this author being removed from the list?

**Comment by 0x6e6562** [comment]
I've just kicked off the build again for [571](https://travis-ci.org/gocql/gocql/builds/47229311), to see if it is an intermittent failure.

**Comment by 0x6e6562** [comment]
OK, so the build passes now and it looks like this patch is in a mergable state - the only thing remaining is to clarify the change to the `AUTHORS` file.

File: AUTHORS (hunk 1 above)
Comment (c0va23) [inline AUTHORS]: It revert-commit.

Comment (0x6e6562) [inline AUTHORS]: Sorry, I don't understand what you mean. @james-maloney submitted #291, so I'm not sure that his name should be removed from the `AUTHORS` list.

Also, in the context of #291, this #287 is effectively undoing the change from #291 - what is the intention behind moving the

```go
if v, ok := value.(Marshaler); ok {
    return v.MarshalCQL(info)
}
```

below the 

```go
if valueRef := reflect.ValueOf(value); valueRef.Kind() == reflect.Ptr {
    if valueRef.IsNil() {
        return nil, nil
    } else if _, ok := value.(Marshaler);  !ok {
        return Marshal(info, valueRef.Elem().Interface())
    }
}
```

block?

Comment (c0va23) [inline AUTHORS]: Without this rollback tests fall. See: https://travis-ci.org/gocql/gocql/builds/47227393

Comment (0x6e6562) [inline AUTHORS]: Right, so you're talking about this test failure:

```
=== RUN TestMarshal
--- FAIL: TestMarshal (0.00 seconds)
panic: value method gocql.CustomString.MarshalCQL called using nil *CustomString pointer [recovered]
```

I can see what you are trying to achieve and in your favor you have written a covering test case, but considering that this is very critical to the library, I think this change requires a second opinion (@phillipCouto).

We need to be sure what the best behavior is.

**Comment by mykaul** [comment]
v2: Added a guard in `DowngradingConsistencyRetryPolicy.Attempt()` that prevents downgrading from SERIAL/LOCAL_SERIAL to a non-serial consistency level (e.g. ONE, QUORUM), which would break serial read (Paxos) guarantees. Also added a unit test verifying this behavior.

**Pull request closed without merging**